### PR TITLE
New: object-shorthand rule (refs: #1617)

### PR DESCRIPTION
**Pull request by xjamundx**
Disallows the following forms:

```js
var foo = {
    x: function() {},
    y: function*() {},
    z: z
};
```

### Todos

- [x]  make this configurable (`"never"`, `"methods"`, `"properties"`, `"always"`)
- [x] Update the docs
- [x] Add a few more tests for more complicated cases

### CODE REVIEW

**Comment by nzakas** [comment]
I like the suggested options. I could go "always" instead of "all", but otherwise, looks good.

File: docs/rules/object-shorthand.md [hunk 1]
```diff
@@ -0,0 +1,46 @@
+# Require Object Literal Shorthand Syntax (object-shorthand)
+
+EcmaScript6 provides a more concise form for defining object literal methods and properties
```
Comment (nzakas) [inline docs/rules/object-shorthand.md]: I'd like to see more of an explanation here, showing the equivalents from ES5.

File: tests/lib/rules/object-shorthand.js [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+    arrowFunctions: true,
+    destructuring: true,
+    generators: true
+};
+
+var eslintTester = new ESLintTester(eslint);
+eslintTester.addRuleTest("lib/rules/object-shorthand", {
+    valid: [
+        { code: "var x = {y() {}}", ecmaFeatures: features },
+        { code: "var x = {y}", ecmaFeatures: features },
```
Comment (nzakas) [inline tests/lib/rules/object-shorthand.js]: You should add some where the property name and value are different, as those should be ignored.

Comment (nzakas) [inline tests/lib/rules/object-shorthand.js]: Also, for both valid and invalid tests, add some with more than two properties.

File: docs/rules/object-shorthand.md [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+including the following forms:
+
+```js
+var x = {
+    go() {},
+    *x() {}
```
Comment (btmills) [inline docs/rules/object-shorthand.md]: Missing a comma.

Now that we have loaders, should we start linting example code in the docs?

Comment (nzakas) [inline docs/rules/object-shorthand.md]: Heh, perhaps we should.

Comment (btmills) [inline docs/rules/object-shorthand.md]: #2271

Comment (xjamundx) [inline docs/rules/object-shorthand.md]: awesome idea!

**Comment by xjamundx** [comment]
Thanks for all the advice. Will get this cleaned up today.

**Comment by xjamundx** [comment]
Added the new options and tried to make the appropriate changes.

**Comment by xjamundx** [comment]
Let me rebase and see what's up....

**Comment by xjamundx** [comment]
I think this is ready, let me know otherwise :)

**Comment by nzakas** [comment]
Lgtm, thanks!

**Pull request merged**